src/chrome/internal/services/sourceToScriptMapper.ts

```diff
@@ -0,0 +1,51 @@
+import { BPRecipeInLoadedSource, BPRecipeInScript } from '../breakpoints/baseMappedBPRecipe';
+import { ConditionalPause, AlwaysPause } from '../breakpoints/bpActionWhenHit';
+import { injectable, inject } from 'inversify';
+import { IBreakpointFeaturesSupport } from '../../cdtpDebuggee/features/cdtpBreakpointFeaturesSupport';
+import { LocationInScript } from '../locations/location';
+import { RangeInScript } from '../locations/rangeInScript';
+import { logger } from 'vscode-debugadapter/lib/logger';
+import { IDebuggeeBreakpointsSetter } from '../../cdtpDebuggee/features/cdtpDebuggeeBreakpointsSetter';
+import { printArray } from '../../collections/printing';
+import { asyncMap } from '../../collections/async';
+import { TYPES } from '../../dependencyInjection.ts/types';
+
+@injectable()
+export class SourceToScriptMapper {
+    private readonly doesTargetSupportColumnBreakpointsCached: Promise<boolean>;
+
+    constructor(
+        @inject(TYPES.IBreakpointFeaturesSupport) private readonly _breakpointFeaturesSupport: IBreakpointFeaturesSupport,
+        @inject(TYPES.IDebuggeeBreakpointsSetter) private readonly _targetBreakpoints: IDebuggeeBreakpointsSetter) {
+        this.doesTargetSupportColumnBreakpointsCached = this._breakpointFeaturesSupport.supportsColumnBreakpoints;
+    }
+
+    public async mapBPRecipe(bpRecipe: BPRecipeInLoadedSource<ConditionalPause | AlwaysPause>): Promise<BPRecipeInScript[]> {
+        const tokensInManyScripts = bpRecipe.location.tokensWhenMappedToScript();
+        return asyncMap(tokensInManyScripts, async manyTokensInScript => {
+            const bestLocation = this.doesTargetSupportColumnBreakpointsCached
+                ? await this.findBestLocationForBP(manyTokensInScript.enclosingRange)
+                : manyTokensInScript.enclosingRange.start; // If we don't support column breakpoints we set the breakpoint at the start of the range
+            const bpRecipeAtBestLocation = new BPRecipeInScript(bpRecipe.unmappedBPRecipe, bestLocation);
+            return bpRecipeAtBestLocation;
+        });
+    }
+
+    private async findBestLocationForBP(range: RangeInScript): Promise<LocationInScript> {
+        const possibleLocations = await this._targetBreakpoints.getPossibleBreakpoints(range);
```

Comment:
I don't know whether this is a real issue or not, but maybe there could be a case where the best breakpoint is outside of the range of mappings? Right now I think we always check getPossibleBreakpoints for the full line, then search for the best one. I guess this would only be an issue if the mapping are missing or incomplete and I can't remember whether there is a specific issue behind this.

But one theoretical example, if you have something like `const foo = bar()` and there is a mapping for `foo = bar()` but not for `const`, but maybe the runtime says that the valid BP location is on `const`... maybe it's something to keep in mind, that the sourcemappings might not give you a broad enough range, but I wouldn't necessarily make any changes right now.

Reply:
I think that is possible, and based on my experience, it'd say it's likely to happen. I'm not sure what is the right way to handle those cases though because we can have minified files where the start of the line is the start of the file, or column breakpoints where if we don't find a breakpoint in the exact place we are looking for, it's difficult to know where we should break.

That's the reason I decided to throw an exception when we don't find a mapping rather than trying to set the breakpoint anyway, so the first time someone finds that case, that someone will file a bug, we'll investigate it, and figure out what is the right thing to do...